### PR TITLE
[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63828
* #63627
* __->__ #63501
* #62861
* #62698

Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

File: torch/quantization/fx/quantization_patterns.py [hunk 1]
```diff
@@ -1456,12 +1459,16 @@ def convert(self,
 @register_quant_pattern(torch.nn.functional.hardtanh)
 @register_quant_pattern(torch.nn.functional.hardtanh_)
 @register_quant_pattern(torch.nn.functional.interpolate)
+@register_quant_pattern(torch.nn.functional.max_pool1d)
```
Comment (raghuramank100) [inline torch/quantization/fx/quantization_patterns.py:1462]: For my understanding, the CopyNodeQuantizeHandler is used for many patterns. How is it that only maxpool1d/2d/3d produces a pattern with a dequant->maxpool->quant? What about other operators like clamp?

File: test/quantization/fx/test_quantize_fx.py [hunk 2]
```diff
@@ -4171,8 +4171,8 @@ def forward(self, x):
         # one quantize_per_tensor for input
         # check exact counts of quantize and dequantize
         count_check = {
-            ns.call_function(torch.quantize_per_tensor) : 1,
-            ns.call_method('dequantize') : 1
+            ns.call_function(torch.quantize_per_tensor) : 2,
```
Comment (raghuramank100) [inline test/quantization/fx/test_quantize_fx.py:4174]: Will a more modular test help? Currently it is not clear what is exactly being tested.

**Review by raghuramank100 — APPROVED** [review]
Looks good, but I think I am missing overall context.

File: torch/quantization/fx/quantization_patterns.py (hunk 1 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:1462]: Yeah I think we should probably remove GeneralTensorValueQuantizeHandler and move everything here, it looks like some backend may not work when we have "quant - some_general_op - dequant", TensorRT included. I think we should just produce "dequant - float_op - quant" pattern for all these ops.
Planning to change that in a separate PR, current goal is to test TensorRT integration.

File: test/quantization/fx/test_quantize_fx.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:4174]: oh, this is testing that we do not have extra quant/dequant for all the ops listed in the forward function, these ops are what we call "CopyNode" in the quantize handler, this test might not be very useful in the future

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

**Pull request closed without merging**